dial.go

```diff
@@ -31,17 +32,31 @@ type DialOptions struct {
 	// KeyFile is the name of the file with the key certificate for TLS. If
 	// specified, CAFile and CrtFile must also be specified. (Optional)
 	KeyFile string
+	// Username is the username of the API user. If specified, APIKey must
+	// also be specified. (Optional)
+	Username string
+	// APIKey is the API key for user authentication. If specified,
+	// Username must also be specified. (Optional)
+	APIKey string
+	// APIUserHeader is the metadata key for the requester's username. This
+	// must be specified if Username and APIKey are specified. (Optional)
+	APIUserHeader string
+	// APIKeyHeader is the metadata key for the requester's API key. This
+	// must be specified if Username and APIKey are specified. (Optional)
+	APIKeyHeader string
 }
 
 func (opts *DialOptions) validate() error {
-	if opts.Address == "" {
-		return errors.New("must provide rpc address")
-	}
-	if opts.TLSConf == nil && (opts.CAFile == "") != (opts.CrtFile == "") != (opts.KeyFile == "") {
-		return errors.New("must provide all or none of the required certificate filenames")
-	}
+	catcher := grip.NewBasicCatcher()
+
+	catcher.AddWhen(opts.Address == "", errors.New("must provide rpc address"))
+	catcher.AddWhen(
+		((opts.CAFile == "") != (opts.CrtFile == "")) || ((opts.CAFile == "") != (opts.KeyFile == "")),
+		errors.New("must provide all or none of the required certificate filenames"),
+	)
+	catcher.AddWhen((opts.Username == "") != (opts.APIKey == ""), errors.New("must provide both a username and API key or neither"))
```

Comment:
This seems like a cute way to get a logical xnor, but I dig it. (no change)

Reply:
I want XORs in go :(